examples/nucleo_g474re/flash_json/main.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2024, Henrik Hose
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <modm/board.hpp>
+#include <modm/debug/logger.hpp>
+#include <nlohmann-json/json.hpp>
+
+#undef MODM_LOG_LEVEL
+#define MODM_LOG_LEVEL modm::log::INFO
+
+// ----------------------------------------------------------------------------
+
+extern "C" const uint32_t __flash_reserved_start[];
+
+using json = nlohmann::json;
+
+constexpr auto max_flash_pages{256};
+
+namespace data
+{
+struct person
+{
+	std::string name;
+	float height;
+	int age;
+	auto
+	operator<=>(const person&) const = default;
+};
+
+void
+to_json(json& j, const person& p)
+{
+	j = json{{"name", p.name}, {"height", p.height}, {"age", p.age}};
+}
+
+void
+from_json(const json& j, person& p)
+{
+	j.at("name").get_to(p.name);
+	j.at("height").get_to(p.height);
+	j.at("age").get_to(p.age);
+}
+}  // namespace data
+
+int
+main()
+{
+	Board::initialize();
+
+	MODM_LOG_INFO << "\n\n+++++++++++++++++++++++++++" << modm::endl;
+	MODM_LOG_INFO << "++ NLOHMANN JSON example ++" << modm::endl;
+	MODM_LOG_INFO << "+++++++++++++++++++++++++++\n" << modm::endl;
+
+	data::person alice_struct = {"Alice", 1.85, 80};
+	json alice_json = alice_struct;
+
+	// basic JSON character string usage
+	{
+		auto alice_json_string = alice_json.dump();
+
+		// imagine that we transmit this string over some interface
+		MODM_LOG_INFO << "JSON from struct is: " << alice_json_string.c_str() << modm::endl;
+
+		// and now we can retrieve the struct from the json object
+		json retrieved_alice_json = json::parse(alice_json_string);
+
+		// JSON is fairly self-annotated, so we might not even need the original struct
+		MODM_LOG_INFO << "\nRetrieved from string JSON has fields ... ";
+		for (auto& [key, value] : retrieved_alice_json.items())
+		{
+			MODM_LOG_INFO << key.c_str() << ", ";
+		}
+		MODM_LOG_INFO << modm::endl;
+		MODM_LOG_INFO.flush();
+
+		// check if retrieved struct is same
+		auto retrieved_alice_struct = retrieved_alice_json.template get<data::person>();
+		if (retrieved_alice_struct == alice_struct)
+		{
+			MODM_LOG_INFO << "\nRetrieved from string and original struct are identical"
+						  << modm::endl;
+		}
+		MODM_LOG_INFO.flush();
+	}
+
+	// basic binary JSON usage
+	{
+		// convert to BSON
+		auto alice_binary = json::to_bson(alice_json);
+
+		MODM_LOG_INFO << "\nBinary JSON (BSON): ";
+		for (const auto& byte : alice_binary) { MODM_LOG_INFO.printf(" 0x%02x", byte); }
+		MODM_LOG_INFO << modm::endl;
+
+		// put BSON in reserved flash
+		uint32_t err{0};
+		const size_t page_start =
```

Comment:
Don't we need to do `+1` to get the first page completely inside the reserved flash area?

Reply:
I think the reserved size in this case is a multiple of the flash area size (`1024*16` in the `project.xml` of the example)?

Reply:
I added an exception in the lbuild module to only allow multiples of 8Kbytes reserved flash (8Kbytes was the largest flash page size I found for some H7 devices). Now reserved flash should always start on its own page. Is this what you had in mind?

Reply:
@rleh ?